### PR TITLE
Add volatile keyword to fields used in double-checked locks

### PR DESCRIPTION
**Pull request by calumgrant**
Across various projects, double-checked locks are implemented without the underlying field being `volatile`. This is correct for the Intel architecture but could fail on other architectures.

Most of these race conditions seem to be benign, but is it still correct to use the `volatile` keyword.

## How were these found
We used lgtm.com to automatically analyse this project. Alerts can be found [here](https://lgtm.com/projects/g/dotnet/corefx/alerts/?mode=tree&ruleFocus=1506088456975).

Disclaimer: I work for lgtm.com (semmle.com).

### CODE REVIEW

File: src/Common/src/System/Data/ProviderBase/DbConnectionPoolGroup.cs [hunk 1]
```diff
@@ -31,7 +31,7 @@ sealed internal class DbConnectionPoolGroup
         private readonly DbConnectionOptions _connectionOptions;
         private readonly DbConnectionPoolKey _poolKey;
         private readonly DbConnectionPoolGroupOptions _poolGroupOptions;
-        private ConcurrentDictionary<DbConnectionPoolIdentity, DbConnectionPool> _poolCollection;
```
Comment (Clockwork-Muse) [inline src/Common/src/System/Data/ProviderBase/DbConnectionPoolGroup.cs]: At least this field isn't set via double-checked locking (it's behind a single explicit lock, or in the constructor), so not relevant.

Comment (calumgrant) [reply src/Common/src/System/Data/ProviderBase/DbConnectionPoolGroup.cs]: The potential race is between line 122 and line 165. But I agree, that `volatile` isn't going to help with the reads on lines 165/170, which is what the query actually found.

Comment (Clockwork-Muse) [reply src/Common/src/System/Data/ProviderBase/DbConnectionPoolGroup.cs]: Ah, you're right, I should have been locking at more of the reads.

File: src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs [hunk 2]
```diff
@@ -56,7 +56,7 @@ public partial class Process : Component
         private bool _priorityBoostEnabled;
         private bool _havePriorityBoostEnabled;
 
-        private bool _raisedOnExited;
+        private volatile bool _raisedOnExited;
```
Comment (Clockwork-Muse) [inline src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs]: This one, especially, would possibly be changeable to some form of `Interlocked.CompareExchange`... except the code is using `lock(this)` (arrghh), which is at least a partial breaking change, and might have wider effects as well.

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs [hunk 3]
```diff
@@ -169,7 +169,7 @@ private class TypeStoreItem : StoreItem
         {
             private readonly object _syncRoot = new object();
             private readonly Type _type;
-            private Dictionary<string, PropertyStoreItem> _propertyStoreItems;
+            private volatile Dictionary<string, PropertyStoreItem> _propertyStoreItems;
```
Comment (Clockwork-Muse) [inline src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: Looks ridiculously amenable to `Lazy<T>`, although I'm not sure of the performance implications.

Comment (calumgrant) [reply src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: `Lazy` is of course the recommended solution, but the whole point of double-checked locking is when you really really need the performance (at the expense of a nest of potential correctness issues.) I didn't want to change the code too much, and of course, such optimisations should be benchmarked but I don't know if such benchmarks have been written.

File: src/Common/src/System/Data/ProviderBase/DbConnectionPoolGroup.cs (hunk 1 above)
Comment (stephentoub) [reply src/Common/src/System/Data/ProviderBase/DbConnectionPoolGroup.cs]: This is going to add a bunch of unnecessary volatile reads and writes, as there are lots of accesses where it's not actually necessary (e.g. all of the reads while holding the lock don't need to be volatile).  I'd rather see this one done with Volatile.Read/Write on just the accesses where it's needed.  This is unlikely to make much of a difference on x86-64, but it could on ARM.

File: src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs (hunk 3 above)
Comment (stephentoub) [reply src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationAttributeStore.cs]: The current access to this looks like this:
```C#
if (_propertyStoreItems == null)
{
    lock (_syncRoot)
    {
        if (_propertyStoreItems == null)
        {
            _propertyStoreItems = CreatePropertyStoreItems();
        }
    }
}

return _propertyStoreItems.TryGetValue(propertyName, out item);
```
It'd be better if it were instead:
```C#
Dictionary<string, PropertyStoreItem> items = _propertyStoreItems;
if (items == null)
{
    lock (_syncRoot)
    {
        items = _propertyStoreItems ?? (_propertyStoreItems = CreatePropertyStoreItems());
    }
}

return items.TryGetValue(propertyName, out item);
```
as in the common case where it's already initialized, that'll avoid an unnecessary secondary volatile read of a field.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs [hunk 4]
```diff
@@ -34,7 +34,7 @@ internal sealed class HttpConnectionPool : IDisposable
 
         private bool _http2Enabled;
         private Http2Connection _http2Connection;
-        private SemaphoreSlim _http2ConnectionCreateLock;
+        private volatile SemaphoreSlim _http2ConnectionCreateLock;
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: This one is also introducing a bunch of unnecessary volatile reads.

I'm not sure this SemaphoreSlim actually needs to be created lazily, but let's assume for now that it is.  We can just delete this code:
```C#
            if (_http2ConnectionCreateLock == null)
            {
                lock (SyncObj)
                {
                    if (_http2ConnectionCreateLock == null)
                    {
                        _http2ConnectionCreateLock = new SemaphoreSlim(1);
                    }
                }
            }
```
and then change the following line:
```C#
await _http2ConnectionCreateLock.WaitAsync().ConfigureAwait(false);
```
to instead be:
```C#
await LazyInitializer.EnsureInitialized(ref _http2ConnectionCreateLock, () => new SemaphoreSlim(1)).WaitAsync().ConfigureAwait(false);
```
and not add volatile to the field.
cc: @geoffkizer

File: src/System.Net.Requests/src/System/Net/FileWebRequest.cs [hunk 5]
```diff
@@ -21,8 +21,8 @@ public class FileWebRequest : WebRequest, ISerializable
         private long _contentLength;
         private int _timeout = DefaultTimeoutMilliseconds;
         private bool _readPending;
-        private bool _writePending;
-        private bool _writing;
+        private volatile bool _writePending;
+        private volatile bool _writing;
```
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/FileWebRequest.cs]: Why does _writing need to be volatile?

File: src/System.Net.Requests/src/System/Net/FtpWebRequest.cs [hunk 6]
```diff
@@ -212,11 +212,11 @@ public sealed class FtpWebRequest : WebRequest
 
         private bool _enableSsl;
         private FtpControlStream _connection;
-        private Stream _stream;
+        private volatile Stream _stream;
```
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/FtpWebRequest.cs]: Why does _stream need to be volatile?  What race condition is this preventing?

File: src/System.Net.Requests/src/System/Net/FtpWebRequest.cs [hunk 7]
```diff
@@ -212,11 +212,11 @@ public sealed class FtpWebRequest : WebRequest
...
         private RequestStage _requestStage;
         private bool _onceFailed;
         private WebHeaderCollection _ftpRequestHeaders;
-        private FtpWebResponse _ftpWebResponse;
+        private volatile FtpWebResponse _ftpWebResponse;
```
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/FtpWebRequest.cs]: Same question

File: src/System.Net.Requests/src/System/Net/FtpWebResponse.cs [hunk 8]
```diff
@@ -16,7 +16,7 @@ public class FtpWebResponse : WebResponse, IDisposable
         private Uri _responseUri;
         private FtpStatusCode _statusCode;
         private string _statusLine;
-        private WebHeaderCollection _ftpRequestHeaders;
+        private volatile WebHeaderCollection _ftpRequestHeaders;
```
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/FtpWebResponse.cs]: I'd prefer to see volatile not added here, and instead just change the Headers property from:
```C#
        public override WebHeaderCollection Headers
        {
            get
            {
                if (_ftpRequestHeaders == null)
                {
                    lock (this)
                    {
                        if (_ftpRequestHeaders == null)
                        {
                            _ftpRequestHeaders = new WebHeaderCollection();
                        }
                    }
                }
                return _ftpRequestHeaders;
            }
        }
```
to instead be:
```C#
        public override WebHeaderCollection Headers =>
            LazyInitializer.EnsureInitialized(ref _ftpRequestHeaders);
```

File: src/System.Net.Requests/src/System/Net/TimerThread.cs [hunk 9]
```diff
@@ -287,7 +287,7 @@ private class InfiniteTimerQueue : Queue
         /// </summary>
         private class TimerNode : Timer
         {
-            private TimerState _timerState;
+            private volatile TimerState _timerState;
```
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/TimerThread.cs]: Why is this needed?  What race condition is it preventing?

File: src/System.Private.Xml/src/System/Xml/Dom/XmlElement.cs [hunk 10]
```diff
@@ -15,7 +15,7 @@ namespace System.Xml
     public class XmlElement : XmlLinkedNode
     {
         private XmlName _name;
-        private XmlAttributeCollection _attributes;
+        private volatile XmlAttributeCollection _attributes;
```
Comment (stephentoub) [inline src/System.Private.Xml/src/System/Xml/Dom/XmlElement.cs]: Same comment as elsewhere that this is adding a bunch of unnecessary volatile reads.  Most accesses are done guarded by HasAttributes, and once we've observed the attributes collection to have been initialized, there's no need to worry about reordering.  If we're going to make this volatile, the access sites should be cleaned up to minimize the number of volatile reads introduced.

File: src/System.Runtime.Caching/src/System/Runtime/Caching/MemoryCache.cs [hunk 11]
```diff
@@ -28,7 +28,7 @@ public class MemoryCache : ObjectCache, IEnumerable, IDisposable
                                                               | DefaultCacheCapabilities.CacheEntryRemovedCallback;
         private static readonly TimeSpan s_oneYear = new TimeSpan(365, 0, 0, 0);
         private static object s_initLock = new object();
-        private static MemoryCache s_defaultCache;
+        private static volatile MemoryCache s_defaultCache;
```
Comment (stephentoub) [inline src/System.Runtime.Caching/src/System/Runtime/Caching/MemoryCache.cs]: Rather than making this volatile, I'd prefer to see the Default property changed to just be:
```C#
public static MemoryCache Default => LazyInitializer.EnsureInitialized(ref s_defaultCache);
```

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/StorePal.cs [hunk 12]
```diff
@@ -17,7 +17,7 @@ namespace Internal.Cryptography.Pal
 {
     internal sealed partial class StorePal
     {
-        private static CollectionBackedStoreProvider s_machineRootStore;
+        private static volatile CollectionBackedStoreProvider s_machineRootStore;
```
Comment (stephentoub) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/StorePal.cs]: Why is this necessary?

File: src/System.Transactions.Local/src/System/Transactions/Transaction.cs [hunk 13]
```diff
@@ -242,7 +242,7 @@ internal Guid DistributedTxId
         // Internal synchronization object for transactions.  It is not safe to lock on the 
         // transaction object because it is public and users of the object may lock it for 
         // other purposes.
-        internal InternalTransaction _internalTransaction;
+        internal volatile InternalTransaction _internalTransaction;
```
Comment (stephentoub) [inline src/System.Transactions.Local/src/System/Transactions/Transaction.cs]: Why is this needed?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs (hunk 4 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs]: Sounds good to me.

File: src/System.Net.Requests/src/System/Net/FileWebRequest.cs (hunk 5 above)
Comment (calumgrant) [reply src/System.Net.Requests/src/System/Net/FileWebRequest.cs]: Due to this result here. https://lgtm.com/projects/g/dotnet/corefx/snapshot/66eb2fbd4851008584c10a12d347ed8fc286bec6/files/src/System.Net.Requests/src/System/Net/FileWebRequest.cs?sort=name&dir=ASC&mode=heatmap#x5eef38a43faf1192:1

File: src/System.Net.Requests/src/System/Net/FtpWebRequest.cs (hunk 6 above)
Comment (calumgrant) [reply src/System.Net.Requests/src/System/Net/FtpWebRequest.cs]: It doesn't - will delete.

File: src/System.Net.Requests/src/System/Net/FileWebRequest.cs (hunk 5 above)
Comment (stephentoub) [reply src/System.Net.Requests/src/System/Net/FileWebRequest.cs]: Thanks, but that doesn't answer my question. What race condition is it preventing?

**Comment by calumgrant** [comment]
Thanks for the review.  I fully hear you that adding `volatile` isn't the desired fix in all cases, and am happy to make the requested changes. I was just seeing `volatile` as the least-risk change. Are you confident that the current tests are sufficient to verify the performance/correctness of these changes?

File: src/System.Net.Requests/src/System/Net/FileWebRequest.cs (hunk 5 above)
Comment (calumgrant) [reply src/System.Net.Requests/src/System/Net/FileWebRequest.cs]: Ha, you are right. `_writing` isn't updated in the critical region.

File: src/System.Net.Requests/src/System/Net/TimerThread.cs (hunk 9 above)
Comment (stephentoub) [reply src/System.Net.Requests/src/System/Net/TimerThread.cs]: Thanks, I see that, but why is the fence supplied by the lock exited immediately after insufficient? This is a primitive value, not an object being published that could be read before fully initialized. Hence my question... what's the actual race condition here being fixed?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/StorePal.cs (hunk 12 above)
Comment (stephentoub) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/StorePal.cs]: Thanks, but that doesn't answer my question. What reordering does this prevent that would otherwise cause a bug?

**Comment by calumgrant** [comment]
Hi there, I'm going to close this, as `volatile` is not strictly needed since `lock` will always insert sufficient memory barriers on modern .Net runtimes. At best it is stylistic. There is a lot of conflicting information about this, where the majority of examples would still use the `volatile` keyword. I do not think there are any real bugs addressed by this fix. Thanks for the consideration.
